src/BinSkim.Rules/SignSecurely.cs

```diff
@@ -0,0 +1,456 @@
+                // configuration file that extends this data).
+                throw new ArgumentException("Unrecognized algorithm id: " + pszObjId, nameof(pszObjId));
+            }
+
+            algorithmId = algorithmData.Name;
+            return algorithmData.Weak;
+        }
+
+        private bool VerifyCodeIsSignedByApprovedCertificate(Native.CERT_INFO certInfo)
+        {
+            // TODO. This code could to ensure that retrieved certificate matches
+            // against an allow list. This functionality will require adding
+            // appropriate options to this rule
+#if NOT_DEFINED
+            // #define SHA256_HASH_LEN 32
+            uint cbKeyId = 32;
+            byte[] rgbKeyId = new byte[cbKeyId];
+            string key;
+
+            if (!Native.CryptHashPublicKeyInfo(
+                IntPtr.Zero,
+                Native.CALG_SHA_256,
+                0,
+                Native.X509_ASN_ENCODING,
+                ref certInfo.SubjectPublicKeyInfo,
+                rgbKeyId,
+                ref cbKeyId
+                ))
+            {
+                validationResult = (ValidationResult)Marshal.GetLastWin32Error();
+                return;
+            }
+
+            // Strip hyphens to match output as provided by tools such as certutil.exe 
+            key = BitConverter.ToString(rgbKeyId).Replace('-', ' ');
+
+            // This call to HashSet.Contains is case-insensitive due to initializing the hashset with an appropriate comparer
+            validationResult = (trustedKeys.Contains(key) ? ValidationResult.Valid : ValidationResult.UnknownCertificateKey);
+#endif
+            return true;
+        }
+
+        private void InvokeCloseAction(Native.WINTRUST_DATA winTrustData)
+        {
+            IntPtr pWinTrustData = IntPtr.Zero;
+            Guid action = Native.ActionGenericVerifyV2;
+            winTrustData.StateAction = Native.StateAction.WTD_STATEACTION_CLOSE;
+
+            try
+            {
+                pWinTrustData = Marshal.AllocHGlobal(Marshal.SizeOf(typeof(Native.WINTRUST_DATA)));
+                Marshal.StructureToPtr(winTrustData, pWinTrustData, false);
+
+                Native.WinVerifyTrust(Native.INVALID_HANDLE_VALUE, ref action, pWinTrustData);
+            }
+            finally
+            {
+                if (pWinTrustData != IntPtr.Zero)
+                {
+                    Marshal.FreeHGlobal(pWinTrustData);
+                }
+                Debug.Assert(winTrustData.File == IntPtr.Zero);
+            }
+        }
+
+        private ValidationResult WinVerifyTrustHelper(IntPtr handle, ref Native.WINTRUST_DATA winTrustData)
+        {
+            Guid action = Native.ActionGenericVerifyV2;
+
+            ValidationResult validationResult;
+            IntPtr pWinTrustData = IntPtr.Zero;
+
+            try
+            {
+                pWinTrustData = Marshal.AllocHGlobal(Marshal.SizeOf(typeof(Native.WINTRUST_DATA)));
```

Comment:
@BillyONeal, I had the same question but stuck with this approach for now, as rewriting for use of 'fixed' statements didn't obviously assist in readability (and since this analysis results in a small number of pinvokes, on the order of the # of binaries under analysis, the perf considerations didn't look urgent). 

@nguerrera has saved the day by recalling to me that proper pinvoke declarations will result in the automatic pinning of memory as required. this will simplify the code considerably. thanks, nick!